### ISSUE TITLE
Improve "Most Popular Products" UX / text on site without `product` post-type.

### ISSUE THREAD

**Issue by jimmymadon** [google/site-kit-wp#7420]
## Feature Description

When a site has the "product" Post Type, we replace the "Most Popular Content" KMW tile with the "Most Popular Products" KMW tile when showing users "answer-based" metrics. However, when a user decides to "Change Metrics", they currently see both of these widget tiles in the Key Metrics Selection panel.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Given that a site does not have a `product` post-type or it did at some point but now does not:
    * Scenario A (most common): The "Most popular metrics" tile is not already picked by the user
        * When a user picks their own metrics, then the metrics selection panel should continue to show the "Most popular content" but hide the "Most popular products" in the list of selectable metrics. 
    * Scenario B (rare edge case): The "Most popular products" metric tile was already selected when the site did have a product post type:
        * Then the "Most popular products" tile should be "checked" and displayed in the Metrics Selection Panel.        
        * The "Most popular products" tile should be rendered in the KMW area on the dashboard and it should display the following message test within: `No product posts currently detected on your site. This metric applies only to sites with product posts.`

## Implementation Brief

The following approach has been tested using the code in the [attached POC PR](https://github.com/google/site-kit-wp/pull/7501).

A. Displaying / hiding the `PopularProducts` widget in the Metrics Selection Panel:
* In order to create a generic way of adding conditions to displaying Key Metrics in the list of available widgets, we should add a `displayInList` property in `assets/js/components/KeyMetrcs/key-metrics-widgets.js` which will define a boolean function to determine if the metric should be displayed in the list or not. 
    * For the `KM_ANALYTICS_POPULAR_PRODUCTS` metric, define this function to accept the `select` function. It should check if this widget has already been picked using the `core-user.isKeyMetricActive()` selector or if the site has any `productBasePaths` defined using the `core-site.getProductBasePaths()` selector. If any of these conditions is true, we should display this widget in the list (return true).
* In `assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js`:
    * When computing the `availableMetrics`, check if `displayInList` is defined and if it is, run it to see if the metric should be filtered out of the list.

B. Showing the `PopularProducts` KMW tile if it has already been selected but there are no product post-types:
* In `assets/js/modules/analytics-4/components/widgets/PopularProductsWidget.js`:
    * Add a check to see if the `KM_ANALYTICS_POPULAR_PRODUCTS` is active using the `core-user.isKeyMetricActive()` selector to the `showWidget` constant and if it is true, then we should show the widget even if there are no `productBasePaths` defined. 
    * Overwrite the `ZeroState` prop message as per the AC iff the site does not have product base paths but if the popular products widget is active (picked already).

### Test Coverage

* No new tests required but the VRTs should be updated.

## QA Brief

* Ensure the WooCommerce plugin is deactivated.
* Setup Site Kit, connect Analytics and turn on the `userInputv2` feature flag so that the Key Metrics set up banner CTA is displayed. Select `Pick your own metrics`. Verify that the `Most popular products by pageviews` widget is not in the list (but all other widgets are). 
* Now activate the WooCommerce plugin. On the dashboard, select `Pick your own metrics` again and verify that the `Most popular products by pageviews` widget is in the list. Select this widget along with one other widget and verify that it displays correctly in the Key Metrics widget area on the dashboard.
* Now deactivate the WooCommerce plugin. Verify that the `Most popular products by pageviews` widget tile is rendered but with the new message as per the AC. Click on the `Change metrics` button in the KMW area and verify that the `Most popular products by pageviews` widget is in the list and it is selected.
* Deselect the `Most popular products by pageviews` widget from the list and select another metric. Ensure the metric is not rendered in the widget area as normal. Now click on `Change metrics` again and ve is rify that the `Most popular products by pageviews` widget is no longer present in the list.

## Changelog entry

* Enhance relevancy of "Most popular products" Key Metric widget availability.

**Comment by marrrmarrr** [comment]
@jimmymadon that suggestion SGTM!

**Comment by techanvil** [comment]
@jimmymadon, I would suggest a tweak to the user flow/AC here.

As described, we'll be showing "Most popular products" as a selectable tile to all of our users, regardless of whether they are, or have any intention of setting up as an ecommerce site.

For what I'd imagine is a pretty substantial proportion of our users, this means we're just showing them an essentially useless item, which takes up space in the selection panel, and I'd imagine is a bit confusing for people who have no thought of enabling products for their site.

Seeing as the rationale behind this is to allow people who have at some point had WooCommerce installed, setup Key Metrics with the "Most popular products" tile enabled, and then disabled/uninstalled WooCommerce, it seems that we're providing a worse user experience for a lot of users to cater for this rare edge case.

What I'd suggest instead is to update the AC such that the "Most popular products" is _not_ shown if there is no `product` post type, _unless_ the tile is already present in the list of key metrics, in which case we do show it so the user can remove the tile.

I don't think this should have a significant impact on the complexity here, but it would be an improved UX.

What do you think?

cc @marrrmarrr

**Comment by techanvil** [comment]
Thanks @jimmymadon, AC LGTM!

**Comment by eugene-manuilov** [comment]
> * When a user picks their own metrics, then the metrics selection panel should continue to show the "Most popular content" but hide the "Most popular products" in the list of selectable metrics.

@jimmymadon @marrrmarrr hiding something when requirements aren't met is not the greatest option in my opinion. People may completely miss it and don't know that that widget even exist because we hide it for them in the selection list.

Let's say they use a custom build products that called subscriptions for their membership driven site. All they need to do to make the most popular products widget work with their subscriptions is to add a custom hook to register their base paths for subscription pages. But to understand that it is available, they need to know that this option it exists.

I think that it would be much better to disable a widget in the selection list if some requirements are not met with a short notice how to activate it. What do you think?

**Comment by eugene-manuilov** [comment]
IB ✔️ though

**Comment by bethanylang** [comment]
@jimmymadon Before we add this is part of the spec for the metrics selection panel redesign, I'd like to get @marrrmarrr's thoughts here. While @eugene-manuilov's note about subscriptions definitely makes sense, we don't often accommodate for use cases that would require adjusting manual code.

Mariya, can you please take a look at the discussion above and let us know your preference here? We can either:
1.  Hide the Most popular products widget altogether if no `products` are found (per @techanvil's comment)
2.  Show the Most popular products widget as "disconnected" (using the pattern that @sigal-teller will develop for #7425) if no `products` are found and provide a notice on how to activate it, which @jimmymadon would need to define

**Comment by eugene-manuilov** [comment]
> While @eugene-manuilov's note about subscriptions definitely makes sense, we don't often accommodate for use cases that would require adjusting manual code.

@bethanylang this is exactly the case where we have an ad-hoc filter to accommodate for edge-cases that will require adjusting manual code. https://github.com/google/site-kit-wp/blob/401cb29a4b8c04e030084684b9cfa29bef6612c1/includes/Core/Assets/Assets.php#L1103-L1112

**Comment by aaemnnosttv** [comment]
Weighing in here with my thoughts –

1. "Products" in terms of this widget was never intended to be specific to WooCommerce, it made sense to use its `product` post type as the default here since it's both generic and will automatically cover the majority of users. The use of a product post type was essentially necessary for us to discover the permalink structure to report on product-specific traffic. In the future, we should use a custom dimension for this to avoid inferences about the content type based on the URL path alone, which is inherently limited in accuracy as it is now.
2. Regarding the filterable base-paths for products, this is indeed not something we would be encouraging our end-users to try and extend on their own. If anything, this would be for other plugins (that provide products for example) to extend, especially since it now involves a list of regular expressions. As mentioned above, I see this specific filter going away in the future in favor of true content-type specific dimension. We'll still want to keep the flexibility to allow for multiple or alternate "product" content types, but this could be exposed in the UI similar to how we did with TwG, potentially also using a filter for extensibility.
3. The most important detail in this issue is that users are not prevented from selecting a different tile if the popular products tile was selected and later "broken" by disabling products
4. The concept of showing all options as disabled where unsatisfied dependencies exist is consistent with how modules are currently represented in the settings (Optimize depends on GA, AdSense requires no ad blocker). This would be out of scope here and should be addressed in #7425

With that said, the current AC seems good to me as a small improvement that we can continue to enhance in the future as we iterate on the design of the metric selection panel. I like the idea of showing the tile as disabled if not selected yet so long as a notice is shown as to why it is disabled. This requires more thought in collaboration with @sigal-teller so I think we should hold on going this far with it in this issue but the idea is great. My only suggestion is that we remove the reference to "post type" in the zero state to instead say something like `No products were detected on your site`.

@marrrmarrr ?

**Comment by marrrmarrr** [comment]
My main concern here is to be clear that this is not something that the user needs to "fix" (they shouldn't feel like something is wrong because no products were detected and they need to add products etc), as I've seen e.g. feedback that some users take our CTAs as "must do" -- e.g. one user thought their dashboard would be "incomplete" if they didn't add AdSense because we have the big CTA in the Monetization section. 

@aaemnnosttv we went back and forth on the copy of this regarding "posts", and even though "products" sounds friendlier, we reverted to posts because that's how we detect this now (a custom dimension would be indeed better in the future as you mentioned), and if the site happens to have some other product-type pages but is not using product posts, then we would be essentially missing that and potentially confusing the user. 

So we could say something along the lines of "No product posts currently detected on your site. This metric applies only to sites with product posts." LMKWYT.
